### ISSUE TITLE
Is it possible to encode recursive types?

### ISSUE THREAD

**Issue by artegoser** [SoftbearStudios/bitcode#35]
![image](https://github.com/user-attachments/assets/02ac48cd-0eba-420c-9543-420dc34f95f0)

![image](https://github.com/user-attachments/assets/b62a6b5c-adad-4122-ae82-8970326c3828)

![image](https://github.com/user-attachments/assets/6606ec54-2f1b-4d25-82c5-1647c4e8bdb8)

In serde version there is error: `type changed`

**Comment by finnbear** [comment]
No, recursive types are not implemented and would be tricky to implement based on how `bitcode@0.6` encodes nested values. You can use `bitcode@0.5` instead, which works very differently and has a `#[bitcode(recursive)]` attribute for `Encode` and native support via `serde`.

**Comment by artegoser** [comment]
Thx. So, is this feature not planned? Or will it be possible again someday?

**Comment by finnbear** [comment]
Not currently planned, but the documentation could be more clear about this so I'll leave the issue open.

**Comment by caibear** [comment]
Can use 0.6 with serde if you remove serde untagged from your [enums](https://github.com/TxtDot/dalet-rs/blob/54c594d3f64176d02525e5e84c41c876cd5e829c/src/daletl.rs#L31). Serde untagged breaks most binary formats even if some don't complain right away.

**Comment by artegoser** [comment]
Yes, it works, only the purpose in reducing the amount of data is not fulfilled. In this case enum identifier is written, which significantly increases the file size. I compared with dlhn, however it cannot decode this. The size becomes larger than my daletpack format, which does not compress well. Is it possible to implement serde(untagged) support in your case?

**Comment by finnbear** [comment]
> Is it possible to implement serde(untagged) support in your case?

No, because `#[serde(untagged)]` works by trying all possible variants until one deserializes. No version of `bitcode` is self-describing, so it's possible that the wrong variant would deserialize without an error but the data would be corrupt.

Making `bitcode` self-describing would probably cost much more data than the enum variants that `#[serde(untagged)]` tries to omit.

The `#[serde(untagged)]` feature is better for textual formats like JSON, where different types have disjoint prefixes like `[` or `{`.

> The size becomes larger than my daletpack format

If you can provide specific details (schema + encoding) of how you can outperform `bitcode` on size, that would be interesting :eyes:

**Comment by artegoser** [comment]
> If you can provide specific details (schema + encoding) of how you can outperform `bitcode` on size, that would be interesting 👀

This format is specifically designed for my schema, but it doesn't compress well (I'm not very knowledgeable about how to optimize this), so I'm looking for solutions that are already out there. It's not for all data types, so it won't help you.

For my format, I had the idea of exposing each type (if it's an enum) and writing a separate identifier for each one.

For example

``rust
enum Enum {
  One(Other),
  Two(Other)
}

enum Other {
  String(String),
  Num(u8)
}
```

Enum in this case has 4 binary identifiers for each type

Enum::OneString
Enum::OneNum

Enum::TwoString
Enum::TwoNum

I don't know how procedurally realizable this is, but it should work. I don't know if this structure compresses well either.

**Comment by caibear** [comment]
After running your [benchmark](https://github.com/TxtDot/dalet-rs/blob/dd597d278c4af03d03207a48ea007e0ab38457c5/examples/bench.rs#L70) it looks like various formats take ~300 bytes. I would recommend testing on a larger dataset if you want to accurately benchmark compression. Compression algorithms are generally inefficient in terms of size and speed on input smaller than tens of kilobytes.

**Comment by DontBreakAlex** [comment]
Hi ! I have a few types that contain `serde_json::Values`s, and I also get a panic with `type changed`. I checked and it seems that serde_json is not using `#[serde(untagged)]`. What's going on ?

**Comment by finnbear** [comment]
I recently added a warning to the README about this; `serde_json::Value` has a custom `Serialize` implementation that is very similar to an untagged enum.

**Comment by artegoser** [comment]
Not only `#[serde(untagged)]`, but also `#[serde(tag = "type")]` and `#[serde(tag = "t", content = "c")]` cause panic: `type changed`

**Comment by finnbear** [comment]
Thanks, I noted this on https://github.com/SoftbearStudios/bitcode/wiki/Serde
